app/receipt.go

```diff
@@ -0,0 +1,177 @@
+package app
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"math/big"
+
+	wasmtypes "github.com/CosmWasm/wasmd/x/wasm/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authsigning "github.com/cosmos/cosmos-sdk/x/auth/signing"
+	"github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/sei-protocol/sei-chain/utils"
+	evmkeeper "github.com/sei-protocol/sei-chain/x/evm/keeper"
+	evmtypes "github.com/sei-protocol/sei-chain/x/evm/types"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+const ShellEVMTxType = math.MaxUint32
+
+var ERC20ApprovalTopic = common.HexToHash("0x8c5be1e5ebec7d5bd14f71427d1e84f3dd0314c0f7b2291e5b200ac8c7c3b925")
+var ERC20TransferTopic = common.HexToHash("0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef")
+var EmptyHash = common.HexToHash("0x0")
+
+type AllowanceResponse struct {
+	Allowance sdk.Int         `json:"allowance"`
+	Expires   json.RawMessage `json:"expires"`
+}
+
+func (app *App) AddCosmosEventsToEVMReceiptIfApplicable(ctx sdk.Context, tx sdk.Tx, checksum [32]byte, response abci.ResponseDeliverTx) {
+	if response.Code > 0 {
+		return
+	}
+	wasmEvents := GetEventsOfType(response, wasmtypes.WasmModuleEventType)
+	logs := []*ethtypes.Log{}
+	for _, wasmEvent := range wasmEvents {
+		contractAddr, found := GetAttributeValue(wasmEvent, wasmtypes.AttributeKeyContractAddr)
+		if !found {
+			continue
+		}
+		// check if there is a ERC20 pointer to contractAddr
+		pointerAddr, _, exists := app.EvmKeeper.GetERC20CW20Pointer(ctx, contractAddr)
+		if exists {
+			action, found := GetAttributeValue(wasmEvent, "action")
+			if !found {
+				continue
+			}
+			var topics []common.Hash
+			switch action {
+			case "mint", "burn", "send", "transfer", "transfer_from", "send_from", "burn_from":
+				topics = []common.Hash{
+					ERC20TransferTopic,
+					app.GetEvmAddressAttribute(ctx, wasmEvent, "from"),
+					app.GetEvmAddressAttribute(ctx, wasmEvent, "to"),
+				}
+				logs = append(logs, &ethtypes.Log{
+					Address: pointerAddr,
+					Index:   uint(len(logs)),
+					Topics:  topics,
+					Data:    common.BigToHash(GetAmountAttribute(wasmEvent)).Bytes(),
+				})
+			case "increase_allowance", "decrease_allowance":
+				ownerStr, found := GetAttributeValue(wasmEvent, "owner")
+				if !found {
+					continue
+				}
+				spenderStr, found := GetAttributeValue(wasmEvent, "spender")
+				if !found {
+					continue
+				}
+				topics := []common.Hash{
+					ERC20ApprovalTopic,
+					app.GetEvmAddressAttribute(ctx, wasmEvent, "owner"),
+					app.GetEvmAddressAttribute(ctx, wasmEvent, "spender"),
+				}
+				res, err := app.WasmKeeper.QuerySmart(
+					ctx,
+					sdk.MustAccAddressFromBech32(contractAddr),
+					[]byte(fmt.Sprintf("{\"allowance\":{\"owner\":\"%s\",\"spender\":\"%s\"}}", ownerStr, spenderStr)),
+				)
+				if err != nil {
+					continue
```

Comment:
is this err something that means a safe 'was not found' situation (or is it worse than that?)  I wonder if it should log?

Reply:
we probably don't want a log here since it simply indicates that the contract does not adhere to cw20 (which is bad from the contract's perspective, but kind of neutral from chain's perspective)